src/waveform/renderers/waveformmark.h

```diff
@@ -94,23 +96,31 @@ class WaveformMark {
     // Check if a point (in image coordinates) lies on drawn image.
     bool contains(QPoint point, Qt::Orientation orientation) const;
 
+    QImage generateImage(float breath, float devicePixelRatio);
+
     QColor m_textColor;
     QString m_text;
     Qt::Alignment m_align;
     QString m_pixmapPath;
 
-    float m_linePosition;
+    float m_linePosition{};
+    int m_breadth{};
 
     WaveformMarkLabel m_label;
 
   private:
     std::unique_ptr<ControlProxy> m_pPositionCO;
     std::unique_ptr<ControlProxy> m_pEndPositionCO;
     std::unique_ptr<ControlProxy> m_pVisibleCO;
-    std::unique_ptr<QOpenGLTexture> m_pTexture; // used by allshader::WaveformRenderMark
+
     friend class allshader::WaveformRenderMark;
+
+    std::unique_ptr<QOpenGLTexture> m_pTexture; // Used by allshader::WaveformRenderMark
+    QImage m_image;                             // Used by WaveformRenderMark
```

Comment:
so IIUC, `m_pTexture` is always `nullptr` when running with a legacy shader and QImage `isNull` the other way around, right? Wouldn't it be better instead then if these concepts lived in the shaders then instead? At least that sounds like thats a better separation of concern.

Reply:
In the renderer you mean? (WaveformMarkRenderer and allshaders::WaveformMarkRenderer). But that would require keeping track of which image or texture goes with which WaveformMark. No, this really needs to be part of WaveformMark. But maybe it can be done by subclassing WaveformMark... I will have a look, I agree this is not ideal.

Reply:
> But that would require keeping track of which image or texture goes with which WaveformMark.

The WaveformMarks in the renderer are just stored in a container, should just be sufficient to store a `std::pair<std::unique_ptr<QOpenGLTexture>, WaveformMark>` and then implement `generateMarkImage` in terms of that, right?

Reply:
No, they are stored in a WaveformMarkSet. But I can derive a WaveformMarkImage and a WaveformMarkTexture from WaveformMark and templatize WaveformMarkSet.

Reply:
I started doing this, but it adds way too much complexity. I prefer having these two data members, of which either one is used depending on the context, instead of adding all this complexity.

Reply:
okay. Have you considered putting the members in a variant to highlight the fact that they are mutual exclusive? If that introduces too much complexity again, then ignore it. I'm just trying to come up with solutions to make the code more self-documenting.

Reply:
yes, that makes sense! will do!

Reply:
done. hope this is in line with what you had in mind.

Reply:
Yes, almost. I was rather thinking of `std::variant` but this also works. The only thing I'm slighly concerned about is the blind static_cast from the parent to the derived type but I don't know any better solution either right now without introducing lots of boilerplate.

Reply:
Thank you for going through the extra effort.